File: dialer/internal/audit/canonicalize.go
package audit

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"sort"
	"strings"
	"time"
)

// sep is the canonical field separator: ASCII Unit Separator 0x1F.
const sep = "\x1f"

// nullSentinel is the two-char literal used for NULL values (MySQL LOAD DATA
// convention). Distinct from the empty string.
const nullSentinel = `\N`

// lpad20 formats a uint64 as a 20-char zero-padded decimal string.
// This ensures the byte-length of numeric fields is stable regardless of
// magnitude, matching MySQL's LPAD(CAST(col AS CHAR), 20, '0').
func lpad20(n uint64) string {
	return fmt.Sprintf("%020d", n)
}

// nullOrStr returns the two-char literal \N for nil pointers; otherwise the
// dereferenced string value.
func nullOrStr(v *string) string {
	if v == nil {
		return nullSentinel
	}
	return *v
}

// nullOrUint returns \N for nil; otherwise the string representation of *v.
func nullOrUint(v *uint64) string {
	if v == nil {
		return nullSentinel
	}
	return fmt.Sprintf("%d", *v)
}

// isoMicros formats a time.Time as ISO 8601 with microseconds and literal Z,
// matching MySQL's DATE_FORMAT(col, '%Y-%m-%dT%H:%i:%s.%fZ').
// The time MUST already be in UTC (enforced by the caller).
func isoMicros(t time.Time) string {
	t = t.UTC()
	return fmt.Sprintf("%04d-%02d-%02dT%02d:%02d:%02d.%06dZ",
		t.Year(), t.Month(), t.Day(),
		t.Hour(), t.Minute(), t.Second(),
		t.Nanosecond()/1000,
	)
}

// nullOrDate formats a nullable time as \N or isoMicros.
func nullOrDate(t *time.Time) string {
	if t == nil {
		return nullSentinel
	}
	return isoMicros(*t)
}

// jcsJSON encodes a value to RFC 8785 canonical JSON (sorted keys, no
// whitespace). Returns \N for nil input.
//
// This matches MySQL's JSON_EXTRACT(col, '$') canonical output for our subset
// of JSON values (no nested arrays of objects with mixed key types).
//
// Limitation: this implementation sorts object keys lexicographically via
// encoding/json's standard sorted-key output. The standard library DOES sort
// object keys in the same order as JCS for the Unicode subset we use (ASCII
// printable + common Latin). Golden fixture tests catch any divergence.
func jcsJSON(v interface{}) string {
	if v == nil {
		return nullSentinel
	}
	b, err := marshalJCS(v)
	if err != nil {
		return nullSentinel
	}
	return string(b)
}

// marshalJCS produces RFC 8785-compliant JSON with sorted object keys.
// For maps, it sorts keys before encoding. For other types, uses
// encoding/json which already produces valid JSON.
func marshalJCS(v interface{}) ([]byte, error) {
	switch val := v.(type) {
	case map[string]interface{}:
		return marshalJCSObject(val)
	case []interface{}:
		return marshalJCSArray(val)
	default:
		// For primitives and struct types, use standard encoding
		return json.Marshal(val)
	}
}

func marshalJCSObject(m map[string]interface{}) ([]byte, error) {
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	var buf strings.Builder
	buf.WriteByte('{')
	for i, k := range keys {
		if i > 0 {
			buf.WriteByte(',')
		}
		keyBytes, err := json.Marshal(k)
		if err != nil {
			return nil, err
		}
		buf.Write(keyBytes)
		buf.WriteByte(':')
		valBytes, err := marshalJCS(m[k])
		if err != nil {
			return nil, err
		}
		buf.Write(valBytes)
	}
	buf.WriteByte('}')
	return []byte(buf.String()), nil
}

func marshalJCSArray(arr []interface{}) ([]byte, error) {
	var buf strings.Builder
	buf.WriteByte('[')
	for i, elem := range arr {
		if i > 0 {
			buf.WriteByte(',')
		}
		b, err := marshalJCS(elem)
		if err != nil {
			return nil, err
		}
		buf.Write(b)
	}
	buf.WriteByte(']')
	return []byte(buf.String()), nil
}

// ---------------------------------------------------------------------------
// Per-table canonical structs
// ---------------------------------------------------------------------------

// AuditLogRow holds the fields needed to build the canonical string for a
// single audit_log row. Matches PLAN §3.5.
type AuditLogRow struct {
	PrevHash    string
	TenantID    uint64
	ID          uint64
	Ts          time.Time
	ActorUserID *uint64
	ActorKind   string
	Action      string
	EntityType  string
	EntityID    *string
	BeforeJSON  interface{}
	AfterJSON   interface{}
	RequestID   *string
	IPAddress   *string
	UserAgent   *string
}

// Canonicalize builds the canonical byte-string for an audit_log row.
// The output is fed to SHA-256 to produce row_hash.
func (r AuditLogRow) Canonicalize() string {
	actorUID := nullSentinel
	if r.ActorUserID != nil {
		actorUID = fmt.Sprintf("%d", *r.ActorUserID)
	}
	fields := []string{
		r.PrevHash,
		lpad20(r.TenantID),
		"audit_log",
		lpad20(r.ID),
		isoMicros(r.Ts),
		actorUID,
		r.ActorKind,
		r.Action,
		r.EntityType,
		nullOrStr(r.EntityID),
		jcsJSON(r.BeforeJSON),
		jcsJSON(r.AfterJSON),
		nullOrStr(r.RequestID),
		nullOrStr(r.IPAddress),
		nullOrStr(r.UserAgent),
	}
	return strings.Join(fields, sep)
}

// Hash computes SHA-256 of the canonical string and returns 64-char lowercase hex.
func (r AuditLogRow) Hash() string {
	return sha256hex(r.Canonicalize())
}

// CallWindowAuditRow holds fields for a call_window_audit row.
type CallWindowAuditRow struct {
	PrevHash         string
	TenantID         uint64
	ID               uint64
	CreatedAt        time.Time
	LeadID           uint64
	PhoneE164        string
	CampaignID       string
	Decision         string
	Reason           string
	TzIana           *string
	TzConfidence     *string
	StateCode        *string
	Zip              *string
	PartyLocal       *time.Time
	PartyDow         *int
	EffectiveOpenMin *int
	EffectiveCloseMin *int
	RuleApplied      *string
	EnforcementPoint string
	NextOpenAt       *time.Time
	CallUUID         *string
}

// Canonicalize builds the canonical byte-string for a call_window_audit row.
func (r CallWindowAuditRow) Canonicalize() string {
	partyDow := nullSentinel
	if r.PartyDow != nil {
		partyDow = fmt.Sprintf("%d", *r.PartyDow)
	}
	effOpen := nullSentinel
	if r.EffectiveOpenMin != nil {
		effOpen = fmt.Sprintf("%d", *r.EffectiveOpenMin)
	}
	effClose := nullSentinel
	if r.EffectiveCloseMin != nil {
		effClose = fmt.Sprintf("%d", *r.EffectiveCloseMin)
	}
	fields := []string{
		r.PrevHash,
		lpad20(r.TenantID),
		"call_window_audit",
		lpad20(r.ID),
		isoMicros(r.CreatedAt),
		fmt.Sprintf("%d", r.LeadID),
		r.PhoneE164,
		r.CampaignID,
		r.Decision,
		r.Reason,
		nullOrStr(r.TzIana),
		nullOrStr(r.TzConfidence),
		nullOrStr(r.StateCode),
		nullOrStr(r.Zip),
		nullOrDate(r.PartyLocal),
		partyDow,
		effOpen,
		effClose,
		nullOrStr(r.RuleApplied),
		r.EnforcementPoint,
		nullOrDate(r.NextOpenAt),
		nullOrStr(r.CallUUID),
	}
	return strings.Join(fields, sep)
}

// Hash returns SHA-256 of the canonical string as 64-char lowercase hex.
func (r CallWindowAuditRow) Hash() string {
	return sha256hex(r.Canonicalize())
}

// ConsentLogRow holds fields for a consent_log row.
type ConsentLogRow struct {
	PrevHash       string
	TenantID       uint64
	ID             uint64
	CallUUID       string
	LeadID         uint64
	PhoneE164      string
	PromptID       string
	DtmfResponse   *string
	Outcome        string
	Language       string
	PromptPlayedAt time.Time
}

// Canonicalize builds the canonical byte-string for a consent_log row.
func (r ConsentLogRow) Canonicalize() string {
	fields := []string{
		r.PrevHash,
		lpad20(r.TenantID),
		"consent_log",
		lpad20(r.ID),
		r.CallUUID,
		fmt.Sprintf("%d", r.LeadID),
		r.PhoneE164,
		r.PromptID,
		nullOrStr(r.DtmfResponse),
		r.Outcome,
		r.Language,
		isoMicros(r.PromptPlayedAt),
	}
	return strings.Join(fields, sep)
}

// Hash returns SHA-256 as 64-char lowercase hex.
func (r ConsentLogRow) Hash() string {
	return sha256hex(r.Canonicalize())
}

// DncSyncLogRow holds fields for a dnc_sync_log row.
type DncSyncLogRow struct {
	PrevHash    string
	ID          uint64
	Source      string
	Kind        string
	FileHash    *string
	Added       int
	Removed     int
	StartedAt   time.Time
	CompletedAt *time.Time
}

// Canonicalize builds the canonical byte-string for a dnc_sync_log row.
// dnc_sync_log is a global table; tenant sentinel = 1.
func (r DncSyncLogRow) Canonicalize() string {
	fields := []string{
		r.PrevHash,
		lpad20(1), // global table sentinel
		"dnc_sync_log",
		lpad20(r.ID),
		r.Source,
		r.Kind,
		nullOrStr(r.FileHash),
		fmt.Sprintf("%d", r.Added),
		fmt.Sprintf("%d", r.Removed),
		isoMicros(r.StartedAt),
		nullOrDate(r.CompletedAt),
	}
	return strings.Join(fields, sep)
}

// Hash returns SHA-256 as 64-char lowercase hex.
func (r DncSyncLogRow) Hash() string {
	return sha256hex(r.Canonicalize())
}

// sha256hex returns the SHA-256 digest of s as 64-char lowercase hex.
func sha256hex(s string) string {
	h := sha256.Sum256([]byte(s))
	return hex.EncodeToString(h[:])
}

File: api/src/services/audit/canonicalize.ts
/**
 * api/src/services/audit/canonicalize.ts
 *
 * Per-row canonical byte-string builder, matching the MySQL BEFORE INSERT
 * trigger exactly (PLAN §3.2).
 *
 * The output is the string that gets SHA-256'd to produce row_hash.
 * Both this module and the Go dialer/internal/audit/canonicalize.go MUST
 * produce the same bytes for every fixture in test/fixtures/canonicalization/.
 *
 * Separator: CHAR(31) = 0x1F (ASCII Unit Separator). Never appears in
 * VARCHAR/JSON columns (Zod enforces in events.ts).
 *
 * NULL serialization: two-char literal '\N' (MySQL LOAD DATA convention).
 * Empty string '' and NULL produce DIFFERENT canonical forms.
 *
 * Timestamps: ISO 8601 with microseconds + literal Z suffix.
 * '%Y-%m-%dT%H:%i:%s.%fZ' in MySQL; toISOStringMicros() here.
 *
 * Numeric columns: LPAD zero to 20 chars so byte-length is stable.
 * JSON columns: RFC 8785 JCS (canonicalize()) — matches MySQL JSON_EXTRACT($).
 */

import { canonicalize as jcs } from '../../../../shared/lib/jcs.js';

export const SEP = '\x1f'; // 0x1F Unit Separator
export const NULL_SENTINEL = '\\N';

export type AuditTable =
  | 'audit_log'
  | 'call_window_audit'
  | 'originate_audit'
  | 'consent_log'
  | 'dnc_sync_log'
  | 'audit_attestation';

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/** Pad a BigInt/number to 20 zero-padded decimal chars. */
export function lpad20(n: bigint | number): string {
  return String(n).padStart(20, '0');
}

/** Serialize a nullable string as '\N' for nulls, raw string otherwise. */
export function nullOrStr(v: string | null | undefined): string {
  return v == null ? NULL_SENTINEL : v;
}

/** Serialize a nullable bigint/number. */
export function nullOrNum(v: bigint | number | null | undefined): string {
  return v == null ? NULL_SENTINEL : String(v);
}

/**
 * Format a Date as ISO 8601 with microseconds + literal Z.
 * MySQL: DATE_FORMAT(col, '%Y-%m-%dT%H:%i:%s.%fZ')
 * Note: JS Date has millisecond precision; we extend to microseconds by
 * appending '000' — matches the DB which stores DATETIME(6).
 */
export function toISOStringMicros(d: Date): string {
  const iso = d.toISOString(); // e.g. 2026-05-12T07:00:00.123Z
  // Replace the trailing 'Z' with '000Z' to get microseconds (6 fractional digits)
  // iso is always ".MMMZ"
  return iso.replace(/\.(\d{3})Z$/, '.$1000Z');
}

/** Nullable date. */
export function nullOrDate(d: Date | null | undefined): string {
  return d == null ? NULL_SENTINEL : toISOStringMicros(d);
}

/** JCS-canonicalize a JSON value; return '\N' for null/undefined. */
export function nullOrJson(v: unknown): string {
  if (v == null) return NULL_SENTINEL;
  return jcs(v);
}

// ---------------------------------------------------------------------------
// Per-table canonical builder types
// ---------------------------------------------------------------------------

export interface AuditLogCanonFields {
  prevHash: string;
  tenantId: bigint;
  id: bigint;
  ts: Date;
  actorUserId: bigint | null;
  actorKind: string;
  action: string;
  entityType: string;
  entityId: string | null;
  beforeJson: unknown;
  afterJson: unknown;
  requestId: string | null;
  ipAddress: string | null;
  userAgent: string | null;
}

export interface CallWindowAuditCanonFields {
  prevHash: string;
  tenantId: bigint;
  id: bigint;
  createdAt: Date;
  leadId: bigint;
  phoneE164: string;
  campaignId: string;
  decision: string;
  reason: string;
  tzIana: string | null;
  tzConfidence: string | null;
  stateCode: string | null;
  zip: string | null;
  partyLocal: Date | null;
  partyDow: number | null;
  effectiveOpenMin: number | null;
  effectiveCloseMin: number | null;
  ruleApplied: string | null;
  enforcementPoint: string;
  nextOpenAt: Date | null;
  callUuid: string | null;
}

export interface OriginateAuditCanonFields {
  prevHash: string;
  tenantId: bigint;
  id: bigint;
  originatedAt: Date;
  leadId: bigint;
  phoneE164: string;
  campaignId: string | null;
  outcome: string;
  tcpaReason: string | null;
  dncDecision: string | null;
  dncSources: unknown;
  tcpaDecision: string | null;
  callUuid: string | null;
}

export interface ConsentLogCanonFields {
  prevHash: string;
  tenantId: bigint;
  id: bigint;
  callUuid: string;
  leadId: bigint;
  phoneE164: string;
  promptId: string;
  dtmfResponse: string | null;
  outcome: string;
  language: string;
  promptPlayedAt: Date;
}

export interface DncSyncLogCanonFields {
  prevHash: string;
  // dnc_sync_log is global; chain uses tenant_id=1 sentinel
  id: bigint;
  source: string;
  kind: string;
  fileHash: string | null;
  added: number;
  removed: number;
  startedAt: Date;
  completedAt: Date | null;
}

// ---------------------------------------------------------------------------
// Canonical string builders
// ---------------------------------------------------------------------------

export function canonicalAuditLog(f: AuditLogCanonFields): string {
  return [
    f.prevHash,
    lpad20(f.tenantId),
    'audit_log',
    lpad20(f.id),
    toISOStringMicros(f.ts),
    f.actorUserId == null ? NULL_SENTINEL : String(f.actorUserId),
    f.actorKind,
    f.action,
    f.entityType,
    nullOrStr(f.entityId),
    nullOrJson(f.beforeJson),
    nullOrJson(f.afterJson),
    nullOrStr(f.requestId),
    nullOrStr(f.ipAddress),
    nullOrStr(f.userAgent),
  ].join(SEP);
}

export function canonicalCallWindowAudit(f: CallWindowAuditCanonFields): string {
  return [
    f.prevHash,
    lpad20(f.tenantId),
    'call_window_audit',
    lpad20(f.id),
    toISOStringMicros(f.createdAt),
    String(f.leadId),
    f.phoneE164,
    f.campaignId,
    f.decision,
    f.reason,
    nullOrStr(f.tzIana),
    nullOrStr(f.tzConfidence),
    nullOrStr(f.stateCode),
    nullOrStr(f.zip),
    nullOrDate(f.partyLocal),
    f.partyDow == null ? NULL_SENTINEL : String(f.partyDow),
    f.effectiveOpenMin == null ? NULL_SENTINEL : String(f.effectiveOpenMin),
    f.effectiveCloseMin == null ? NULL_SENTINEL : String(f.effectiveCloseMin),
    nullOrStr(f.ruleApplied),
    f.enforcementPoint,
    nullOrDate(f.nextOpenAt),
    nullOrStr(f.callUuid),
  ].join(SEP);
}

export function canonicalOriginateAudit(f: OriginateAuditCanonFields): string {
  return [
    f.prevHash,
    lpad20(f.tenantId),
    'originate_audit',
    lpad20(f.id),
    toISOStringMicros(f.originatedAt),
    String(f.leadId),
    f.phoneE164,
    nullOrStr(f.campaignId),
    f.outcome,
    nullOrStr(f.tcpaReason),
    nullOrStr(f.dncDecision),
    nullOrJson(f.dncSources),
    nullOrStr(f.tcpaDecision),
    nullOrStr(f.callUuid),
    nullOrJson(f.dncSources), // payload field — per PLAN §3.5
  ].join(SEP);
}

export function canonicalConsentLog(f: ConsentLogCanonFields): string {
  return [
    f.prevHash,
    lpad20(f.tenantId),
    'consent_log',
    lpad20(f.id),
    f.callUuid,
    String(f.leadId),
    f.phoneE164,
    f.promptId,
    nullOrStr(f.dtmfResponse),
    f.outcome,
    f.language,
    toISOStringMicros(f.promptPlayedAt),
  ].join(SEP);
}

export function canonicalDncSyncLog(f: DncSyncLogCanonFields): string {
  return [
    f.prevHash,
    lpad20(1), // global table; tenant sentinel = 1
    'dnc_sync_log',
    lpad20(f.id),
    f.source,
    f.kind,
    nullOrStr(f.fileHash),
    String(f.added),
    String(f.removed),
    toISOStringMicros(f.startedAt),
    nullOrDate(f.completedAt),
  ].join(SEP);
}

/** Compute SHA-256 over the canonical string. Returns lowercase 64-char hex. */
export async function hashCanonical(canonical: string): Promise<string> {
  const { createHash } = await import('node:crypto');
  return createHash('sha256').update(canonical, 'utf8').digest('hex');
}
